agent/engine/dockerstate/docker_task_engine_state.go

```diff
@@ -258,6 +272,79 @@ func (state *DockerTaskEngineState) RemoveENIAttachment(mac string) {
 	}
 }
 
+// AllEBSAttachments returns all the ebs volumes managed by ecs on the instance
+func (state *DockerTaskEngineState) AllEBSAttachments() []*apiebs.ResourceAttachment {
+	state.lock.RLock()
+	defer state.lock.RUnlock()
+
+	return state.allEBSAttachmentsUnsafe()
+}
+
+func (state *DockerTaskEngineState) allEBSAttachmentsUnsafe() []*apiebs.ResourceAttachment {
+	var allEBSAttachments []*apiebs.ResourceAttachment
+	for _, v := range state.ebsAttachments {
+		allEBSAttachments = append(allEBSAttachments, v)
+	}
+	return allEBSAttachments
+}
+
+// AllPendingEBSAttachments returns all the ebs volumes managed by ecs on the instance that haven't sent a state change yet
+func (state *DockerTaskEngineState) AllPendingEBSAttachments() []*apiebs.ResourceAttachment {
+	state.lock.RLock()
+	defer state.lock.RUnlock()
+
+	return state.allPendingEBSAttachmentsUnsafe()
+}
+
+func (state *DockerTaskEngineState) allPendingEBSAttachmentsUnsafe() []*apiebs.ResourceAttachment {
+	var allEBSAttachments []*apiebs.ResourceAttachment
+	for _, v := range state.ebsAttachments {
+		if !v.IsSent() {
+			allEBSAttachments = append(allEBSAttachments, v)
+		}
+	}
+	return allEBSAttachments
+}
+
+// AddEBSAttachment adds the ebs volume to state
+func (state *DockerTaskEngineState) AddEBSAttachment(ebsAttachment *apiebs.ResourceAttachment) {
+	if ebsAttachment == nil {
+		seelog.Debug("Cannot add empty ebs attachment information")
+		return
+	}
+	state.lock.Lock()
+	defer state.lock.Unlock()
+	if _, ok := state.ebsAttachments[ebsAttachment.AttachmentProperties[apiebs.VolumeIdName]]; !ok {
```

Comment:
nit: i think we can break this down for better readability

iiuc, you're checking if state.ebsAttachments map has a key with the new attachment's volume id?

Reply:
yep, that's a good call out